res/controllers/Pioneer-DDJ-400-script.js

```diff
@@ -0,0 +1,774 @@
+            engine.setValue(group, "hotcue_"+hotcuePad+"_set", 1);
+        }
+
+        this.keyboardModeRefCount[channel] = 0; // reset count
+        this.keyboardModeEnabledOutput(channel, group);
+        return;
+    }
+
+    // if hotcue point is set perform coresponding halftone operation
+    if (value > 0) {
+        // count pressed Pad per deck
+        this.keyboardModeRefCount[channel] += 1;
+        var newValue = this.halftoneToPadMap[padNum-1];
+
+        engine.setValue(group, "pitch", newValue);
+        engine.setValue(group, "hotcue_"+hotcuePad+"_gotoandplay", 1);
+    } else {
+        // decrease the number of active Pads, this should minimize unwanted stops
+        this.keyboardModeRefCount[channel] -= 1;
+        if (this.keyboardModeRefCount[channel] <= 0) {
+            engine.setValue(group, "hotcue_"+hotcuePad+"_gotoandstop", 1);
+            engine.setValue(group, "pitch", 0.0); // reset pitch
+            this.keyboardModeRefCount[channel] = 0; // reset refcount to 0
+        }
+    }
+};
+
+PioneerDDJ400.keyshiftModePad = function(_channel, control, value, _status, group) {
+    "use strict";
+    if (value === 0) {
+        return; // ignore release
+    }
+    engine.setValue(group, "pitch", this.halftoneToPadMap[control & 0xf]);
+};
+
+PioneerDDJ400.samplerModeShiftPadPressed = function(_channel, _control, value, _status, group) {
+    "use strict";
+    if (value === 0) {
+        return; // ignore release
+    }
+
+    var playing = engine.getValue(group, "play");
+    // when playing stop and return to start/cue point
+    if (playing > 0) {
+        engine.setValue(group, "cue_gotoandstop", 1);
+    } else { // load selected track
+        engine.setValue(group, "LoadSelectedTrack", 1);
+    }
+};
+
+
+PioneerDDJ400.shiftPressed = function(channel, _control, value) {
+    "use strict";
+    this.shiftState[channel] = value;
+};
+
+PioneerDDJ400.waveFormRotate = function(_channel, _control, value) {
+    "use strict";
+    // select the Waveform to zoom left shift = deck1, right shift = deck2
+    var deckNum = this.shiftState[0] > 0 ? 1 : 2;
+    var oldVal = engine.getValue("[Channel"+deckNum+"]", "waveform_zoom");
+    var newVal = oldVal + (value > 0x64 ? 1 : -1);
+
+    engine.setValue("[Channel"+deckNum+"]", "waveform_zoom", newVal);
+};
+
+PioneerDDJ400.loopin4beatPressed = function(channel, _control, value, _status, group) {
+    "use strict";
+    var loopEnabled = engine.getValue(group, "loop_enabled");
+    this.loopin4beat[channel] = (value > 0);
+
+    if (loopEnabled === 0 && value > 0) {
+        engine.setValue(group, "loop_in", 1);
+    }
+};
+
+PioneerDDJ400.loopin4beatPressedLong = function(_channel, _control, value, _status, group) {
+    // problematic - loop gets set to the playback position where the 'long press' was recognized
+    // and not to the point at which the button was initially pressed
+    // as a result, the loop is not set where one would expect
```

Comment:
You could just use the position when the button is released. Not ideal, but I don't have a better idea.

Reply:
Ehh, I don't think that would be practically useful either.

Reply:
I'm not sure if there's an easy solution to this one - in rekordBox, the software is somehow remembering the playback position at which the button is initially pressed, and then once it determines that the button was held long enough to trigger a "long press" then it knows to use that position (instead of the current playback position) as the beginning of a 4-beat loop.

The current implementation of loopin4beatPressedLong simply sets beatloop_4_activate at the current playback position, which if course has moved by the duration of the "long press delay" by the time this happens.

It may be most expedient to just remove this functionality until and unless it can be addressed.

Reply:
I think as with the comment below, to implement this properly requires extending the control interface to allow a script to more precisely set a loop. So I'll proceed to remove it (with a note in the comments about it).

Reply:
This is removed. Marking as resolved.